packages/performance/src/services/perf_logger.test.ts

```diff
@@ -108,12 +113,131 @@ describe('Performance Monitoring > perf_logger', () => {
+        EXPECTED_TRACE_MESSAGE
+      );
+    });
+
+    it('Amount of customAttribute is 5.', () => {
+      const EXPECTED_TRACE_MESSAGE =
+        `{` +
+        webAppInfo +
+        `,"trace_metric":{"name":"${TRACE_NAME}","is_auto":false,\
+"client_start_time_us":${START_TIME * 1000},"duration_us":${DURATION * 1000},\
+"custom_attributes":{"attr1":"val1","attr2":"val2","attr3":"val3","attr4":"val4","attr5":"val5"}}}`;
+      getIidStub.returns(IID);
+      stub(attributeUtils, 'getVisibilityState').returns(VISIBILITY_STATE);
+      SettingsService.getInstance().loggingEnabled = true;
+      SettingsService.getInstance().logTraceAfterSampling = true;
+      const trace = new Trace(TRACE_NAME);
+      for (let i = 1; i <= 5; i++) {
+        trace.putAttribute('attr' + i, 'val' + i);
+      }
+      trace.record(START_TIME, DURATION);
+      clock.tick(1);
+
+      expect(addToQueueStub).to.be.called;
+      expect(addToQueueStub.getCall(0).args[0].message).to.be.equal(
+        EXPECTED_TRACE_MESSAGE
+      );
+    });
+  });
+
+  describe('logPageLoadTrace', () => {
+    it('creates, serializes and sends a page load trace to cc service', () => {
+      const flooredStartTime = Math.floor(TIME_ORIGIN * 1000);
+      const EXPECTED_TRACE_MESSAGE = `{"application_info":{"google_app_id":"${APP_ID}",\
+"app_instance_id":"${IID}","web_app_info":{"sdk_version":"${SDK_VERSION}",\
+"page_url":"${PAGE_URL}","service_worker_status":${SERVICE_WORKER_STATUS},\
+"visibility_state":${attributeUtils.VisibilityState.VISIBLE},"effective_connection_type":${EFFECTIVE_CONNECTION_TYPE}},\
+"application_process_state":0},"trace_metric":{"name":"_wt_${PAGE_URL}","is_auto":true,\
+"client_start_time_us":${flooredStartTime},"duration_us":${DURATION * 1000},\
+"counters":{"domInteractive":10000,"domContentLoadedEventEnd":20000,"loadEventEnd":10000,\
+"_fp":40000,"_fcp":50000,"_fid":90000}}}`;
+      getIidStub.returns(IID);
+      SettingsService.getInstance().loggingEnabled = true;
+      SettingsService.getInstance().logTraceAfterSampling = true;
+
+      stub(attributeUtils, 'getVisibilityState').returns(attributeUtils.VisibilityState.VISIBLE);
+
+      const navigationTiming: PerformanceNavigationTiming = {
+        domComplete: 100,
+        domContentLoadedEventEnd: 20,
+        domContentLoadedEventStart: 10,
+        domInteractive: 10,
+        loadEventEnd: 10,
+        loadEventStart: 10,
+        redirectCount: 10,
+        type: 'navigate',
+        unloadEventEnd: 10,
+        unloadEventStart: 10,
+        duration: DURATION,
+      } as PerformanceNavigationTiming;
+
+      const navigationTimings: PerformanceNavigationTiming[] = [
+        navigationTiming
+      ];
+
+      const firstPaint: PerformanceEntry = {
+        name: 'first-paint',
+        startTime: 40,
+        duration: 100,
+        entryType: 'url',
+        toJSON() {},
+      };
+
+      const firstContentfulPaint:Readonly<PerformanceEntry> = {
```

Comment:
Why is it `Readonly<PerformanceEntry>` here but it is `PerformanceEntry` for `firstPaint` above?

Reply:
Fixed it. Classic Copy Paste issue!